File: packages/go/provision.go
package s3tests

import (
	"bytes"
	"context"
	"errors"
	"fmt"

	"github.com/aws/smithy-go"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	s3types "github.com/aws/aws-sdk-go-v2/service/s3/types"
	s3vectors "github.com/cloud-portable/s3vectors/packages/go"
)

// BucketInfo is a provisioned $bucket prerequisite; Name feeds
// ${res.<handle>.name}.
type BucketInfo struct {
	Name string
	// KnownKeys are object keys the runner observed being written to this
	// bucket. Teardown deletes them explicitly before the listing-based
	// sweep: some servers' listings miss keys (e.g. MinIO hides "foo/bar"
	// when object "foo" exists), which would otherwise leak the bucket.
	KnownKeys []string
}

// ObjectInfo is a provisioned $object prerequisite; the fields feed
// ${res.<handle>.key|etag|versionId}.
type ObjectInfo struct {
	Key       string
	ETag      string
	VersionID string
}

// Provisioner establishes vector prerequisites and tears their resources
// down. Implementations may target the server's own API (see
// DefaultProvisioner) or any out-of-band mechanism.
type Provisioner interface {
	// Bucket provisions a bucket named name per the prerequisite spec
	// (versioning "": unset, "Enabled" or "Suspended"; ObjectLock enabled
	// at creation). The runner chooses name.
	Bucket(ctx context.Context, t Target, p *s3vectors.BucketPrerequisite, name string) (BucketInfo, error)
	// Object seeds bucketName with an object. body holds the resolved
	// content-descriptor bytes (nil when the prerequisite declares none —
	// still create the object, empty).
	Object(ctx context.Context, t Target, p *s3vectors.ObjectPrerequisite, bucketName string, body []byte) (ObjectInfo, error)
	// Teardown removes the vector's buckets and their contents,
	// best-effort. Returned errors become VectorResult.Warnings, never
	// failures (e.g. COMPLIANCE-retained objects are legitimately
	// undeletable).
	Teardown(ctx context.Context, t Target, buckets []BucketInfo) []error
}

// DefaultProvisioner provisions prerequisites against the endpoint under
// test itself, using the main identity.
type DefaultProvisioner struct{}

var _ Provisioner = DefaultProvisioner{}

func (DefaultProvisioner) Bucket(ctx context.Context, t Target, p *s3vectors.BucketPrerequisite, name string) (BucketInfo, error) {
	in := &s3.CreateBucketInput{Bucket: aws.String(name)}
	if p.ObjectLock != nil && *p.ObjectLock {
		in.ObjectLockEnabledForBucket = aws.Bool(true)
	}
	if t.Region != "us-east-1" {
		in.CreateBucketConfiguration = &s3types.CreateBucketConfiguration{
			LocationConstraint: s3types.BucketLocationConstraint(t.Region),
		}
	}
	if _, err := t.Client.CreateBucket(ctx, in); err != nil {
		return BucketInfo{}, fmt.Errorf("CreateBucket %s: %w", name, err)
	}
	if p.Versioning != "" {
		_, err := t.Client.PutBucketVersioning(ctx, &s3.PutBucketVersioningInput{
			Bucket: aws.String(name),
			VersioningConfiguration: &s3types.VersioningConfiguration{
				Status: s3types.BucketVersioningStatus(p.Versioning),
			},
		})
		if err != nil {
			return BucketInfo{}, fmt.Errorf("PutBucketVersioning %s=%s: %w", name, p.Versioning, err)
		}
	}
	return BucketInfo{Name: name}, nil
}

func (DefaultProvisioner) Object(ctx context.Context, t Target, p *s3vectors.ObjectPrerequisite, bucketName string, body []byte) (ObjectInfo, error) {
	in := &s3.PutObjectInput{
		Bucket: aws.String(bucketName),
		Key:    aws.String(p.Key),
		Body:   bytes.NewReader(body),
	}
	if p.ContentType != "" {
		in.ContentType = aws.String(p.ContentType)
	}
	if len(p.Metadata) > 0 {
		in.Metadata = p.Metadata
	}
	out, err := t.Client.PutObject(ctx, in)
	if err != nil {
		return ObjectInfo{}, fmt.Errorf("PutObject %s/%s: %w", bucketName, p.Key, err)
	}
	return ObjectInfo{
		Key:       p.Key,
		ETag:      aws.ToString(out.ETag),
		VersionID: aws.ToString(out.VersionId),
	}, nil
}

func (DefaultProvisioner) Teardown(ctx context.Context, t Target, buckets []BucketInfo) []error {
	var errs []error
	for _, b := range buckets {
		errs = append(errs, teardownBucket(ctx, t.Client, b.Name, b.KnownKeys)...)
	}
	return errs
}

// teardownBucket empties and deletes one bucket, best-effort: abort multipart
// uploads, delete every version and delete marker (bypassing governance
// retention and lifting legal holds on object-lock buckets), then delete the
// bucket itself.
func teardownBucket(ctx context.Context, client *s3.Client, bucket string, knownKeys []string) []error {
	var errs []error

	// Abort in-flight multipart uploads.
	var keyMarker, uploadIDMarker *string
	for {
		mu, err := client.ListMultipartUploads(ctx, &s3.ListMultipartUploadsInput{
			Bucket: aws.String(bucket), KeyMarker: keyMarker, UploadIdMarker: uploadIDMarker,
		})
		if isNoSuchBucket(err) {
			return nil // already gone (e.g. a step deleted it) — nothing to do
		}
		if err != nil {
			errs = append(errs, fmt.Errorf("teardown %s: ListMultipartUploads: %w", bucket, err))
			break
		}
		for _, u := range mu.Uploads {
			if _, err := client.AbortMultipartUpload(ctx, &s3.AbortMultipartUploadInput{
				Bucket: aws.String(bucket), Key: u.Key, UploadId: u.UploadId,
			}); err != nil {
				errs = append(errs, fmt.Errorf("teardown %s: AbortMultipartUpload %s: %w", bucket, aws.ToString(u.Key), err))
			}
		}
		if mu.IsTruncated == nil || !*mu.IsTruncated {
			break
		}
		keyMarker, uploadIDMarker = mu.NextKeyMarker, mu.NextUploadIdMarker
	}

	// Delete keys the runner knows it wrote, in case the server's listings
	// miss them (best-effort; the sweep below reports anything that fails).
	for _, key := range knownKeys {
		client.DeleteObject(ctx, &s3.DeleteObjectInput{Bucket: aws.String(bucket), Key: aws.String(key)})
	}

	// AWS rejects BypassGovernanceRetention on buckets without object lock,
	// so only send it (and lift legal holds) where lock is configured.
	locked := bucketHasObjectLock(ctx, client, bucket)

	var keyM, verM *string
	for {
		lv, err := client.ListObjectVersions(ctx, &s3.ListObjectVersionsInput{
			Bucket: aws.String(bucket), KeyMarker: keyM, VersionIdMarker: verM,
		})
		if err != nil {
			errs = append(errs, fmt.Errorf("teardown %s: ListObjectVersions: %w", bucket, err))
			break
		}
		var ids []s3types.ObjectIdentifier
		for _, v := range lv.Versions {
			if locked {
				// Best-effort: legal holds block deletion even with bypass.
				client.PutObjectLegalHold(ctx, &s3.PutObjectLegalHoldInput{
					Bucket: aws.String(bucket), Key: v.Key, VersionId: v.VersionId,
					LegalHold: &s3types.ObjectLockLegalHold{Status: s3types.ObjectLockLegalHoldStatusOff},
				})
			}
			ids = append(ids, s3types.ObjectIdentifier{Key: v.Key, VersionId: v.VersionId})
		}
		for _, m := range lv.DeleteMarkers {
			ids = append(ids, s3types.ObjectIdentifier{Key: m.Key, VersionId: m.VersionId})
		}
		for chunk := range slicesChunk(ids, 1000) {
			in := &s3.DeleteObjectsInput{
				Bucket: aws.String(bucket),
				Delete: &s3types.Delete{Objects: chunk, Quiet: aws.Bool(true)},
			}
			if locked {
				in.BypassGovernanceRetention = aws.Bool(true)
			}
			out, err := client.DeleteObjects(ctx, in)
			if err != nil {
				errs = append(errs, fmt.Errorf("teardown %s: DeleteObjects: %w", bucket, err))
				continue
			}
			for _, e := range out.Errors {
				errs = append(errs, fmt.Errorf("teardown %s: delete %s (%s): %s",
					bucket, aws.ToString(e.Key), aws.ToString(e.VersionId), aws.ToString(e.Message)))
			}
		}
		if lv.IsTruncated == nil || !*lv.IsTruncated {
			break
		}
		keyM, verM = lv.NextKeyMarker, lv.NextVersionIdMarker
	}

	if _, err := client.DeleteBucket(ctx, &s3.DeleteBucketInput{Bucket: aws.String(bucket)}); err != nil && !isNoSuchBucket(err) {
		errs = append(errs, fmt.Errorf("teardown %s: DeleteBucket: %w", bucket, err))
	}
	return errs
}

func isNoSuchBucket(err error) bool {
	if err == nil {
		return false
	}
	var ae smithy.APIError
	return errors.As(err, &ae) && (ae.ErrorCode() == "NoSuchBucket" || ae.ErrorCode() == "NotFound")
}

func bucketHasObjectLock(ctx context.Context, client *s3.Client, bucket string) bool {
	out, err := client.GetObjectLockConfiguration(ctx, &s3.GetObjectLockConfigurationInput{Bucket: aws.String(bucket)})
	if err != nil {
		// Any error (typically ObjectLockConfigurationNotFoundError) means:
		// don't send lock-specific teardown requests.
		return false
	}
	return out.ObjectLockConfiguration != nil &&
		out.ObjectLockConfiguration.ObjectLockEnabled == s3types.ObjectLockEnabledEnabled
}

// slicesChunk yields s in chunks of at most n.
func slicesChunk(s []s3types.ObjectIdentifier, n int) func(func([]s3types.ObjectIdentifier) bool) {
	return func(yield func([]s3types.ObjectIdentifier) bool) {
		for len(s) > 0 {
			m := min(n, len(s))
			if !yield(s[:m]) {
				return
			}
			s = s[m:]
		}
	}
}

File: packages/js/lib/provision.js
// Prerequisite provisioning and best-effort teardown against the endpoint
// under test itself (the default provisioner), using the main identity.

import {
  CreateBucketCommand, PutBucketVersioningCommand, PutObjectCommand,
  ListMultipartUploadsCommand, AbortMultipartUploadCommand,
  GetObjectLockConfigurationCommand, ListObjectVersionsCommand,
  PutObjectLegalHoldCommand, DeleteObjectsCommand, DeleteObjectCommand,
  DeleteBucketCommand
} from '@aws-sdk/client-s3'

/** The built-in Provisioner: CreateBucket/PutObject via the runner's client. */
export const defaultProvisioner = {
  async bucket (target, prereq, name, { signal } = {}) {
    const input = { Bucket: name }
    if (prereq.objectLock === true) input.ObjectLockEnabledForBucket = true
    if (target.region !== 'us-east-1') {
      input.CreateBucketConfiguration = { LocationConstraint: target.region }
    }
    try {
      await target.client.send(new CreateBucketCommand(input), { abortSignal: signal })
    } catch (err) {
      throw new Error(`CreateBucket ${name}: ${err.message}`)
    }
    if (prereq.versioning) {
      try {
        await target.client.send(new PutBucketVersioningCommand({
          Bucket: name,
          VersioningConfiguration: { Status: prereq.versioning }
        }), { abortSignal: signal })
      } catch (err) {
        throw new Error(`PutBucketVersioning ${name}=${prereq.versioning}: ${err.message}`)
      }
    }
    return { name, knownKeys: [] }
  },

  async object (target, prereq, bucketName, body, { signal } = {}) {
    const input = {
      Bucket: bucketName,
      Key: prereq.key,
      Body: body ?? new Uint8Array(0)
    }
    if (prereq.contentType) input.ContentType = prereq.contentType
    if (prereq.metadata && Object.keys(prereq.metadata).length > 0) input.Metadata = prereq.metadata
    let out
    try {
      out = await target.client.send(new PutObjectCommand(input), { abortSignal: signal })
    } catch (err) {
      throw new Error(`PutObject ${bucketName}/${prereq.key}: ${err.message}`)
    }
    return { key: prereq.key, etag: out.ETag ?? '', versionId: out.VersionId ?? '' }
  },

  async teardown (target, buckets, { signal } = {}) {
    const warnings = []
    for (const b of buckets) {
      warnings.push(...await teardownBucket(target.client, b.name, b.knownKeys ?? [], signal))
    }
    return warnings
  }
}

const isNoSuchBucket = (err) => err?.name === 'NoSuchBucket' || err?.name === 'NotFound'

// teardownBucket empties and deletes one bucket, best-effort: abort multipart
// uploads, delete known-written keys explicitly (some servers' listings hide
// keys), delete every version and delete marker (bypassing governance
// retention and lifting legal holds on object-lock buckets), then delete the
// bucket.
async function teardownBucket (client, bucket, knownKeys, signal) {
  const warnings = []
  const send = (cmd) => client.send(cmd, { abortSignal: signal })

  // Abort in-flight multipart uploads.
  try {
    let keyMarker, uploadIdMarker
    do {
      const mu = await send(new ListMultipartUploadsCommand({
        Bucket: bucket, KeyMarker: keyMarker, UploadIdMarker: uploadIdMarker
      }))
      for (const u of mu.Uploads ?? []) {
        try {
          await send(new AbortMultipartUploadCommand({ Bucket: bucket, Key: u.Key, UploadId: u.UploadId }))
        } catch (err) {
          warnings.push(`teardown ${bucket}: AbortMultipartUpload ${u.Key}: ${err.message}`)
        }
      }
      keyMarker = mu.IsTruncated ? mu.NextKeyMarker : undefined
      uploadIdMarker = mu.IsTruncated ? mu.NextUploadIdMarker : undefined
    } while (keyMarker !== undefined)
  } catch (err) {
    if (isNoSuchBucket(err)) return warnings // already gone — nothing to do
    warnings.push(`teardown ${bucket}: ListMultipartUploads: ${err.message}`)
  }

  // Delete keys the runner knows it wrote, in case the server's listings
  // miss them (best-effort; the sweep below reports anything that fails).
  for (const key of knownKeys) {
    try {
      await send(new DeleteObjectCommand({ Bucket: bucket, Key: key }))
    } catch { /* the sweep reports residuals */ }
  }

  // AWS rejects BypassGovernanceRetention on buckets without object lock,
  // so only send it (and lift legal holds) where lock is configured.
  const locked = await bucketHasObjectLock(client, bucket, signal)

  try {
    let keyMarker, versionIdMarker
    do {
      const lv = await send(new ListObjectVersionsCommand({
        Bucket: bucket, KeyMarker: keyMarker, VersionIdMarker: versionIdMarker
      }))
      const ids = []
      for (const v of lv.Versions ?? []) {
        if (locked) {
          try {
            await send(new PutObjectLegalHoldCommand({
              Bucket: bucket, Key: v.Key, VersionId: v.VersionId, LegalHold: { Status: 'OFF' }
            }))
          } catch { /* legal holds block deletion even with bypass */ }
        }
        ids.push({ Key: v.Key, VersionId: v.VersionId })
      }
      for (const m of lv.DeleteMarkers ?? []) ids.push({ Key: m.Key, VersionId: m.VersionId })
      for (let i = 0; i < ids.length; i += 1000) {
        const input = { Bucket: bucket, Delete: { Objects: ids.slice(i, i + 1000), Quiet: true } }
        if (locked) input.BypassGovernanceRetention = true
        try {
          const out = await send(new DeleteObjectsCommand(input))
          for (const e of out.Errors ?? []) {
            warnings.push(`teardown ${bucket}: delete ${e.Key} (${e.VersionId ?? ''}): ${e.Message}`)
          }
        } catch (err) {
          warnings.push(`teardown ${bucket}: DeleteObjects: ${err.message}`)
        }
      }
      keyMarker = lv.IsTruncated ? lv.NextKeyMarker : undefined
      versionIdMarker = lv.IsTruncated ? lv.NextVersionIdMarker : undefined
    } while (keyMarker !== undefined)
  } catch (err) {
    warnings.push(`teardown ${bucket}: ListObjectVersions: ${err.message}`)
  }

  try {
    await send(new DeleteBucketCommand({ Bucket: bucket }))
  } catch (err) {
    if (!isNoSuchBucket(err)) warnings.push(`teardown ${bucket}: DeleteBucket: ${err.message}`)
  }
  return warnings
}

async function bucketHasObjectLock (client, bucket, signal) {
  try {
    const out = await client.send(new GetObjectLockConfigurationCommand({ Bucket: bucket }), { abortSignal: signal })
    return out.ObjectLockConfiguration?.ObjectLockEnabled === 'Enabled'
  } catch {
    return false // typically ObjectLockConfigurationNotFoundError
  }
}
